### ISSUE TITLE
Feature Request: Generalized Linear Models

### ISSUE THREAD

**Issue by anath2** [abstractqqq/polars_ds_extension#338]
Hi! Thanks for working on this. Are there any plans to support GLM such as [poisson regression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.PoissonRegressor.html#poissonregressor) in the future?

**Comment by abstractqqq** [comment]
Hey this is interesting.

Sklearn relies on Scipy's lgfbs solver, which is mentioned in this wiki [article](https://en.wikipedia.org/wiki/Limited-memory_BFGS).

To work properly with Polars, I will have to find a Rust impl of lgfbs solver, which might be hard to find since the existing implementations are not widely used. This might take a very long time and I might have to do a lot of research on my own.

**Comment by anath2** [comment]
Thanks for the pointers. I will also try to work on it in my spare time and perhaps open a PR

**Comment by abstractqqq** [comment]
Here are some resources I found but I lack the time fully digest and implement..

GLM: https://en.wikipedia.org/wiki/Generalized_linear_model
IRLS: http://www.nicksun.fun/statistics/2023/03/01/IRLS-from-scratch.html
Algorithm: https://us.sagepub.com/sites/default/files/upm-binaries/21121_Chapter_15.pdf

**Comment by alipatti** [comment]
I recently implemented GLMs in rust (using faer) a few weeks ago for another project and I'm very interested to see this functionality come to Polars. Mind if I take a stab?

**Comment by abstractqqq** [comment]
Go ahead! That will be huge for the project!

**Comment by anath2** [comment]
@alipatti Hi - I've added some code to start implementing this and a list of things to do [here](https://github.com/abstractqqq/polars_ds_extension/pull/351). Feel free to work any aspect of this.

**Comment by abstractqqq** [comment]
I merged the PR and I was reading about it last night. Good work! I might make an attempt on the limited memory version

**Comment by anath2** [comment]
Great! Thanks. 
I'm thinking I might reorganize `linalg` a bit after the finishing the implementation. 
Maybe something like:
 ```
linalg
      |-------------glm
                 |--- mod.rs
                 |--- link_functions.rs
                 |--- glm_family.rs
                 |--- solvers.rs
      
      |-------------lr
                |--- mod.rs
                |--- solvers.rs
      
      |-------------online_lr
               |--- mod.rs
               |--- solvers.rs
      
      |---mod.rs

      |---[shared stuff]
      |---        ....

```
What do you think?

**Comment by abstractqqq** [comment]
It makes sense to separate like this because GLMs require more constructs like link function, variance, etc..

Btw I am reviewing your code and I have 2 comments.

1. In faer_irls, when checking for convergences, we use the full beta even even when has_bias = true. In elastic net, I don't use change in bias as part of the max norm.. Can you double check.

2. I found some faer optimizations and ways to reuse allocated memory. I might make a PR with these updates soon.

**Comment by anath2** [comment]
> In faer_irls, when checking for convergences, we use the full beta even even when has_bias = true. In elastic net, I don't use change in bias as part of the max norm.. Can you double check.

Yes indeed. In `faer_coordinate_descent` you loop over only the coefficients whereas here we use the full beta. Why is that?  Should we do the same in `faer_irls`? 

Btw I looked at your draft PR. Thanks! for putting time into this. I'll start working on the TODOS after tomorrow.

**Comment by abstractqqq** [comment]
> > In faer_irls, when checking for convergences, we use the full beta even even when has_bias = true. In elastic net, I don't use change in bias as part of the max norm.. Can you double check.
> 
> Yes indeed. In `faer_coordinate_descent` you loop over only the coefficients whereas here we use the full beta. Why is that? Should we do the same in `faer_irls`?
> 
> Btw I looked at your draft PR. Thanks! for putting time into this. I'll start working on the TODOS after tomorrow.

I think it is reasonable to argue that the contribution from the intercept shouldn't be treated the same as the contribution from the main coefficients in any descent algorithms. E.g in gradient descent derivative w.r.t the constant term isn't the same as derivatives with respect to the coefficients.. 

In IRLS, we are not using derivatives of beta, but derivatives of mu, and then it makes sense to all entries in the beta vec the same.. That's my opinion. It is probably not a big deal anyways.
